File: app/src/main/java/quickbeer/android/viewmodels/BeerListViewModel.java
/**
 * This file is part of QuickBeer.
 * Copyright (C) 2016 Antti Poikela <antti.poikela@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package quickbeer.android.viewmodels;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.List;

import io.reark.reark.data.DataStreamNotification;
import ix.Ix;
import quickbeer.android.data.DataLayer;
import quickbeer.android.data.pojos.ItemList;
import quickbeer.android.providers.ProgressStatusProvider;
import quickbeer.android.utils.StringUtils;
import rx.Observable;
import rx.functions.Func1;
import rx.observables.ConnectableObservable;
import rx.schedulers.Schedulers;
import rx.subjects.PublishSubject;
import rx.subscriptions.CompositeSubscription;
import timber.log.Timber;

import static io.reark.reark.utils.Preconditions.get;

public abstract class BeerListViewModel extends NetworkViewModel<ItemList<String>> {

    @NonNull
    private final DataLayer.GetBeer getBeer;

    @NonNull
    private final DataLayer.GetBeerSearch getBeerSearch;

    @NonNull
    private final SearchViewViewModel searchViewViewModel;

    @NonNull
    private final ProgressStatusProvider progressStatusProvider;

    @NonNull
    private final PublishSubject<Observable<DataStreamNotification<ItemList<String>>>> source = PublishSubject.create();

    @NonNull
    private final PublishSubject<List<BeerViewModel>> beers = PublishSubject.create();

    protected BeerListViewModel(@NonNull DataLayer.GetBeer getBeer,
                                @NonNull DataLayer.GetBeerSearch getBeerSearch,
                                @NonNull SearchViewViewModel searchViewViewModel,
                                @NonNull ProgressStatusProvider progressStatusProvider) {
        this.getBeer = get(getBeer);
        this.getBeerSearch = get(getBeerSearch);
        this.searchViewViewModel = get(searchViewViewModel);
        this.progressStatusProvider = get(progressStatusProvider);
    }

    @NonNull
    protected abstract Observable<DataStreamNotification<ItemList<String>>> dataSource();

    @NonNull
    protected abstract Observable<DataStreamNotification<ItemList<String>>> reloadSource();

    protected boolean reportsProgress() {
        return true;
    }

    @NonNull
    public Observable<List<BeerViewModel>> getBeers() {
        return beers.asObservable();
    }

    public void reload() {
        source.onNext(reloadSource());
    }

    @Override
    protected void bind(@NonNull CompositeSubscription subscription) {
        // Switch source according to selector subject
        ConnectableObservable<DataStreamNotification<ItemList<String>>> sharedObservable =
                Observable.switchOnNext(source)
                        .subscribeOn(Schedulers.computation())
                        .publish();

        // Construct progress status. Completed with value means we'll receive onNext.
        subscription.add(sharedObservable
                .filter(notification -> !notification.isCompletedWithValue())
                .map(toProgressStatus())
                .startWith(ProgressStatus.LOADING)
                .distinctUntilChanged()
                .subscribe(this::setProgressStatus));

        // Actual update
        subscription.add(sharedObservable
                .filter(DataStreamNotification::isOnNext)
                .map(DataStreamNotification::getValue)
                .doOnNext(beerSearch -> Timber.d("Search finished"))
                .map(ItemList::getItems)
                .map(toBeerViewModelList())
                .doOnNext(list -> Timber.d("Publishing " + list.size() + " beers"))
                .subscribe(beers::onNext));

        // Share progress status to progress provider
        if (reportsProgress()) {
            subscription.add(progressStatusProvider
                    .addProgressObservable(sharedObservable
                    .map(notification -> notification)));
        }

        subscription.add(sharedObservable.connect());

        // Start with default data source
        source.onNext(dataSource());

        // Switch to search on new search term
        subscription.add(get(searchViewViewModel)
                .getQueryStream()
                .distinctUntilChanged()
                .filter(StringUtils::hasValue)
                .doOnNext(query -> Timber.d("query(%s)", query))
                .map(query -> get(getBeerSearch).call(query))
                .subscribe(source::onNext, Timber::e));
    }

    @NonNull
    private Func1<List<Integer>, List<BeerViewModel>> toBeerViewModelList() {
        return beerIds -> Ix.from(beerIds)
                .map(id -> new BeerViewModel(id, getBeer, progressStatusProvider))
                .toList();
    }

    @Override
    protected boolean hasValue(@Nullable ItemList<String> item) {
        return !get(item).getItems().isEmpty();
    }
}

File: app/src/main/java/quickbeer/android/viewmodels/BeerListViewModel.kt
/**
 * This file is part of QuickBeer.
 * Copyright (C) 2017 Antti Poikela <antti.poikela@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package quickbeer.android.viewmodels

import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject
import io.reark.reark.data.DataStreamNotification
import quickbeer.android.data.actions.BeerActions
import quickbeer.android.data.actions.BeerSearchActions
import quickbeer.android.data.pojos.ItemList
import quickbeer.android.providers.ProgressStatusProvider
import quickbeer.android.utils.kotlin.hasValue
import timber.log.Timber

abstract class BeerListViewModel
protected constructor(
    private val beerActions: BeerActions,
    private val beerSearchActions: BeerSearchActions,
    private val searchViewViewModel: SearchViewViewModel,
    private val progressStatusProvider: ProgressStatusProvider
) : NetworkViewModel<ItemList<String>>() {

    private val source = BehaviorSubject.create<Observable<DataStreamNotification<ItemList<String>>>>()

    private val beers = PublishSubject.create<List<BeerViewModel>>()

    protected abstract fun dataSource(): Observable<DataStreamNotification<ItemList<String>>>

    protected abstract fun reloadSource(): Observable<DataStreamNotification<ItemList<String>>>

    protected open fun reportsProgress(): Boolean {
        return true
    }

    open fun isRefreshable(): Boolean {
        return true
    }

    fun getBeers(): Observable<List<BeerViewModel>> {
        return beers.hide()
    }

    fun reload() {
        Timber.v("reload()")
        source.onNext(reloadSource())
    }

    override fun bind(disposable: CompositeDisposable) {
        Timber.v("bind()")

        // Start with default data source
        source.onNext(dataSource())

        // Switch source according to selector subject
        val sharedObservable =
            Observable.switchOnNext(source)
                .subscribeOn(Schedulers.computation())
                .publish()

        // Construct progress status. Completed with value means we'll receive onNext.
        disposable.add(sharedObservable
            .filter { !it.isCompletedWithValue }
            .map(toProgressStatus())
            .startWith(NetworkViewModel.ProgressStatus.LOADING)
            .distinctUntilChanged()
            .doOnNext { Timber.v("New progress status: $it") }
            .subscribe({ setProgressStatus(it) }, Timber::e))

        // Actual update
        disposable.add(sharedObservable
            .filter { it.isOnNext }
            .map { it.value!! }
            .doOnNext { Timber.d("Search finished") }
            .map { it.items.map { BeerViewModel(it, false, beerActions, progressStatusProvider) } }
            .doOnNext { Timber.d("Publishing %s beers", it.size) }
            .subscribe({ beers.onNext(it) }, Timber::e))

        // Share progress status to progress provider
        if (reportsProgress()) {
            disposable.add(
                progressStatusProvider
                    .addProgressObservable(sharedObservable.map { it }))
        }

        disposable.add(sharedObservable.connect())

        // Switch to search on new search term
        disposable.add(searchViewViewModel
            .getQueryStream()
            .distinctUntilChanged()
            .filter { it.hasValue() }
            .doOnNext {
                Timber.d("query(%s)", it)
                Timber.v("Emptying list for new query")
                beers.onNext(emptyList())
            }
            .map { beerSearchActions.search(it) }
            .subscribe({ source.onNext(it) }, Timber::e))
    }
}
